Postgres: ability to create indexes

arzavj opened the pull request:
resolves #804 

### Description

Rough first pass at addressing the issue. Would love to get some thoughts on whether I'm on the right track. Would also love to get some help on adding a unit and integration test so that I can make sure that this code works. 

I'm imaging that usage will look like:
```
{{ config(
    indexes=[
      { 'columns': ['column_a', 'column_b'], 'unique': false },
      { 'columns': ['column_a', 'column_c'], 'unique': true },
    ]
) }}
```

This PR gives developers the ability to:
- create multiple indexes inside the same transaction as the `create_table_as`
- create a mix of unique and non-unique indexes
- create indexes on any number of columns
- address the issue where indexes aren't recreated (as discussed in this [discourse](https://discourse.getdbt.com/t/issue-with-post-hooks-not-running-with-dbt-0-16-1/1235/6)) by prepending the index name with a timestamp; I choose to prepend instead of append because if the column names are long or there are a number of columns in the index then postgres might truncate the index name and drop the timestamp
- address the issue where sometimes only one index out of a few indexes is created because the indexes share the same first few columns -- for example, `['column_a', 'column_b']` and `['column_a', 'column_c']`; what happens is that postgres truncates the final index name to fit a certain max length; so in this example postgres might drop the second column from the index name causing both index names to be the same (and hence only one of them is created); of course certain conditions need to be met in order to run into this problem (i.e. shared initial columns and possibly long column names)

Note that this does not support gin indexes and indexes on expressions of columns like `lower(column_a)`
Also note that I'm assuming snakecased column and table names.

### Checklist
 - [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
 - [x] I have run this code in development and it appears to resolve the stated issue
 - [x] This PR includes tests, or tests are not required/relevant for this PR
 - [x] I have updated the `CHANGELOG.md` and added information about my change to the "dbt next" section.

brylie commented on plugins/postgres/dbt/include/postgres/macros/adapters.sql:
```diff
@@ -14,6 +14,44 @@
+  create {% if unique -%}
+    unique
+  {%- endif %} index if not exists
+  "{{ index_name }}" on {{ relation }}
```
In some cases, it is desirable to specify the type of index, via `USING`. E.g.

```sql
CREATE INDEX gin_idx ON documents_table USING GIN (locations);
```

jtcohen6 reviewed (commented):
Nice start here @arzavj, and thanks for taking so much into consideration for the first go!

Let's figure out the right handoff points between `create_indexes` and the materialization code. I don't think it makes sense to pass back a semicolon-delimited string and append it to `post_hooks`. I'm drawing inspiration from `persist_docs`, which feels like clear, compelling, and quite similar functionality.

Once we're settled on the syntax, we'll definitely need to add tests for the various permutations.

> Also note that I'm assuming snakecased column and table names.

I think you're good! The model name, if quoted and weirdly cased, may "just work" because of how dbt prints `{{ relation }}`. If the columns are weirdly quoted/cased, the user will have to account for that when defining the `indexes` config.

arzavj commented:
@jtcohen6 please take a look at the updated PR based on the comments above. Some open questions I have:
1. You mentioned "_avoid any operation unless the model both has an index defined and is running on an adapter with an implementation of create_indexes_". Is the approach I've take correct? How does one check to see whether an adapter has implemented `get_create_index_sql`?
1. Should I call `do create_indexes` before or after the `run_hooks(post_hooks, inside_transaction=True)` call?
1. You mentioned updating the incremental materialization as well. Do we also need to take care of the seed and snapshot materializations?
1. Should I just call `config.get('indexes')` as I do now or create a method on the config like `config.persist_relation_docs()`?
1. If I do the previous point, I could add validation to ensure that `type` is one of `[btree, hash, gist, gin]`. Do you recommend that or is it just better to let the query run and possibly fail with an invalid type?
1. I'm not too sure on the purpose of `call statement`. Is it meant to track calls and their results? For what purpose? Should I be calling `call statement` outside the loop to track the whole `create_indexes` call or inside loop with a different name for each index to track the individual calls?

arzavj commented on plugins/postgres/dbt/include/postgres/macros/adapters.sql:
```diff
@@ -14,6 +14,25 @@
+  {#- only be created on every other run. See -#}
+  {#- https://github.com/fishtown-analytics/dbt/issues/1945#issuecomment-576714925 for -#}
+  {#- an explanation. -#}
+  {%- set index_name = md5("_".join([run_started_at.strftime("%s"), relation.identifier] + index_dict['columns'])) -%}
```
I'm not sure if this `md5` call will work since I don't understand how the dbt core utils are included and made available in `.sql` files.

arzavj commented on core/dbt/include/global_project/macros/adapters/common.sql:
```diff
@@ -51,6 +51,29 @@
+{% macro default__create_indexes(relation) -%}
+  {%- set _indexes = config.get('indexes', default=[]) -%}
+
+  {%- call statement('create_indexes') -%}
```
I'm not too sure on the purpose of `call statement`. Is it meant to track calls and their results? For what purpose? Should I be calling `call statement` outside the loop to track the whole `create_indexes` call or inside loop with a different name for each index to track the individual calls?

arzavj commented on core/dbt/include/global_project/macros/adapters/common.sql:
```diff
@@ -51,6 +51,29 @@
+{%- endmacro %}
+
+{% macro default__create_indexes(relation) -%}
+  {%- set _indexes = config.get('indexes', default=[]) -%}
```
Should I just call `config.get('indexes')` as I do now or create a method on the config like `config.persist_relation_docs()`?

jtcohen6 replied:
I think `persist_relation_docs` just exists to validate that `persist_docs` is a dict. We can do that in other ways, e.g. via methods to the Postgres class object in python (see other comments).

arzavj commented on core/dbt/include/global_project/macros/materializations/table/table.sql:
```diff
@@ -47,6 +47,8 @@
 
   {{ adapter.rename_relation(intermediate_relation, target_relation) }}
 
+  {% do create_indexes(target_relation) %}
```
Should I call `do create_indexes` before or after the `run_hooks(post_hooks, inside_transaction=True)` call?

jtcohen6 commented on core/dbt/include/global_project/macros/adapters/common.sql:
```diff
@@ -51,6 +51,29 @@
+{% endmacro %}
+
+{% macro default__get_create_index_sql(relation, index_dict) -%}
+  {{ exceptions.raise_not_implemented(
+    'get_create_index_sql macro not implemented for adapter '+adapter.type()) }}
```
Let's just have this return `None`, rather than raise an exception. Then, `create_indexes` can run the query if and only if the SQL returned is not `None`. So by default, nothing happens; on Postgres, the appropriate SQL is returned and run.

jtcohen6 reviewed (commented):
Nice work @arzavj, this is really coming along!

> 1. You mentioned "avoid any operation unless the model both has an index defined and is running on an adapter with an implementation of create_indexes". Is the approach I've take correct? How does one check to see whether an adapter has implemented get_create_index_sql?

If an adapter has implemented `get_create_index_sql`, dbt will use its version; otherwise, dbt will use `default__get_create_index_sql`.

As a feature, indexes are unique to Postgres among the "core four" databases, but they're pretty common out in the world of databases. That's why I think we should implement `create_indexes` in such a way that:
- By default, it does nothing
- At the same time, any database-agnostic plumbing is reusable (= included in the default implementation)
- Any PostgreSQL-specific components are wrapped in the `postgres__` implementation

The best way to accomplish this, to my mind, was two dispatch macros:
- **`create_indexes`**: Called by materialization. Loops over indexes, calls `get_create_index_sql` for each. If SQL is returned, call `run_query()` for it; if nothing is returned, do nothing. All of this can be part of `default__create_indexes`; it's unlikely someone would need to change/override this, but you never know.
- **`get_create_index_sql`**: Called by `create_indexes`, one index at a time. Returns the database-specific SQL for creating that index. `default__get_create_index_sql` should return `None` (most databases don't have indexes), instead of raising an exception (what you have now). `postgres__get_create_index_sql` should return the appropriate PostgreSQL (exactly as you have it now). An Oracle, SQLServer, or Materialize user out in the world could reimplement _just_ `get_create_index_sql` with the right SQL, and still benefit from all of the default plumbing in `create_indexes`. (They'd also need to define a custom `IndexConfig` in the python portion of the adapter, but that's a different story.)

> 2. You mentioned updating the incremental materialization as well. Do we also need to take care of the seed and snapshot materializations?

Yes!! Really good point, thank you for catching that. The [`seed` materialization](https://github.com/fishtown-analytics/dbt/blob/develop/core/dbt/include/global_project/macros/materializations/seed/seed.sql) logic will look a lot like the incremental (I think `{% if full_refresh_mode or not exists_as_table %}`). I think the [`snapshot` materialization](https://github.com/fishtown-analytics/dbt/blob/develop/core/dbt/include/global_project/macros/materializations/snapshot/snapshot.sql#L185) logic can be even simpler (`{% if not target_relation_exists %}`), since there's no such thing as full-refreshing.

> 3. If I do the previous point, I could add validation to ensure that type is one of [btree, hash, gist, gin]. Do you recommend that or is it just better to let the query run and possibly fail with an invalid type?

See comment below. I'll fully admit I'm not a Postgres pro; depending on how many index types there are, how customizable they are, and how frequently they change, it may make sense to validate in python (set in stone), in Jinja (standardized but can be overridden), or not at all.

arzavj commented on plugins/postgres/dbt/adapters/postgres/impl.py:
```diff
@@ -16,6 +16,7 @@
 @dataclass
 class PostgresConfig(AdapterConfig):
     unlogged: Optional[bool] = None
+    indexes: Optional[List[Dict[str, Any]]] = None
```
That makes sense. I'll go with leaving it as is and not doing any validation since it will mean less to keep in sync between dbt code and postgres docs and newer versions. The error shown for an incorrect type will be thrown by postgres and should make sense to the user.

arzavj commented:
@jtcohen6 I've responded to all the comments and made sure that the existing tests still pass! I think it's time to add tests for the code added in this PR. Are there specific examples of integration and unit tests that you think I could use for inspiration as I write these?

jtcohen6 commented:
Brilliant, the code is looking really good!

I think a new integration test is appropriate, `065_postgres_tests/test_indexes.py`, that features:
- a table model with indexes (run once)
- an incremental model with indexes (run, run again, run `--full-refresh`)
- a seed with indexes (seed, seed again, seed `--full-refresh`)
- a snapshot with indexes (snapshot twice)

Each of those can be something silly, `select 1 as fun` or similar, as long as the defined indexes are valid.

And maybe also some failure cases:
- Invalid index config (assert it raises the exceptions you've defined)
- Index with a non-existing type (assert it fails / returns a database error)

arzavj commented:
Apologies for the delay on this @jtcohen6; I'm on vacation and will be back on the 29th. Will get to this then!

arzavj commented:
@jtcohen6 Please take a look at the tests I added. I also updated the changelog. I haven't added any documentation or comments. Please let me know if I should.

arzavj commented:
A couple more things:
1. The unit tests are [failing](https://dev.azure.com/fishtown-analytics/dbt/_build/results?buildId=3181&view=logs&j=1c778e8e-596b-5a0a-22c2-460364725d7c&t=1607d490-6a74-5e3f-97a9-72eb94e82d47&l=249) however when I run the same command `python -m pytest -v test/unit` locally they pass (with 1 skip). Weirdly though when I run `make test` I get `ERROR: InterpreterNotFound: python3.8`; not sure why it's trying to use python 3.8 when I'm on python 3.7.10.
2. I'm struggling to figure out why the cla-bot is unable to recognize me. My git config has the correct email address as output by `git config --list | grep email`. This matches my primary Github email address.

jtcohen6 commented:
@arzavj The tests are looking great! I haven't had a chance to do a deep dive, but I'm really happy to see everything you've done here.

Quick things:
1. That Windows unit test has been failing on all PRs, and it's on us to fix it (#3233). In the meantime, we just merged a PR that disabled it for now, so if you pull from `develop` it should be skipped on the next one.
2. I'm not positive, but think the issue is that one of your commits (2e5ff4036866370dccac43f6f001444aa70734a8) is co-authored by your GitHub account, and by... also your GitHub account? I'm not sure:

![Screen Shot 2021-04-12 at 11 05 27 AM](https://user-images.githubusercontent.com/13897643/114417999-18da2200-9b80-11eb-9d18-f8f578209fde.png)

arzavj commented:
You were right! Fixed the author of that culprit commit and also rebased on the latest develop.

arzavj commented:
hey @jtcohen6! Bumping this up on your radar :)

jtcohen6 reviewed (approved):
@arzavj Sorry for the delay!

I love how this feature "just works," from the unique/type properties down to the guaranteed-unique index name. And the tests look _amazing_. Thank you so much for this contribution, and for all your thorough work here over the past few months!

This will be released in v0.20.0 :)

arzavj commented:
Thank you for the kind words @jtcohen6! Really appreciate all your help on this PR! Excited to update my code to use this new feature once v0.20.0 is released :)
